### DISCUSSION TITLE
Testing spack package changes to see if it will break builds on a host of machines/environments

### DISCUSSION THREAD

**Discussion by psakievich** [spack/spack#26334] [Ideas]
# Motivation
As a I've become more dependent on spack I find that I am wanting a way to test changes in spack against builds that I am responsible for to anticipate and guard against changes.  Spack core features are relatively well tested, but changes to package variants and conflicts often give me a host of headaches when I update spack.  What is really frustrating about this is that I have no way to anticipate all the problems I am going to have. I just have to update spack, have lots of things break across machines and then try to fix them as fast as possible.  This is not really sustainable so it has to be done infrequently, but that also makes it a much bigger burden to do and there is not a good way to anticipate the magnitude of difficulty I will encounter.

# New feature proposal
As a user I would like a way to concretize a given `spack.yaml` file (or some other representative file for an environment) and then diff the resulting `spack.lock` against a reference `spack.lock` with a range of warnings and errors output.  
I would like a warning for hash changes to signify a likely rebuild of a package on the machine
I would like errors for concretization failures and an explanation of why the error occurred: new conflict, removed/added variant etc.

To make this scalable I would like to be able to perform this concretization on a different machine so I can test changes for a large collection of environments and machines in one location through a CI process.  This means that the concretization process needs to be able to ignore the local machine's architecture and just run with the one I supply in the `spack.yaml` or some other representative file.  

In order for this to work the `spack.yaml` would need to be a complete representation of the machine and configs, but not a complete representation of the concrete environment like `spack.lock`.  So it will also likely be necessary to add a feature create a such a representative environment file from a less verbose `spack.yaml` that is currently being used. Maybe something like 
```
spack -e . config generate complete -n machine_specific_env.yaml
```
Where we are naming the generated `spack.yaml` `machine_specific_env.yaml`

# Proposed workflow

On a machine/environment you'd like to start testing spack changes against:
```
# create some environment and navigate there
spack -e .  concretize
spack -e . config generate complete -n complete_env.yaml
# copy complete_env.yaml.yaml and the spack.lock files to a reference git repo
cp complete_env.yaml [some ref git repo]/machine_name/
cp spack.lock [some ref git repo]/machine_name/ref_spack.lock
git commit -a -m "Add config for env on machine_name"
git push
```

on a testing machine
```
spack cd -r
git pull origin develop
# navigate to machine_name directory inside git repo
spack env create -d test complete_env.yaml
spack concretize -reference_machine
spack env lock-diff spack.lock test/spack.lock
```

This is a little clunky as written but I wanted to be verbose to show the steps.  I imagine we could combine steps and just allow an option to concretize directly without needing to create an environment.  Possible the last 3 lines could all be combined into something like:
`spack concretize --yaml complete_env.yaml --diff spack.lock`

**Comment by alalazo** [comment]
What would you say are the core differences between what you propose here and the Gitlab Pipelines that are run on each PR? 

For reference here's an example of [a pipeline](https://gitlab.spack.io/spack/spack/-/pipelines/66656) referring to a [commit](https://github.com/spack/spack/commit/c5410bd3337f2c7645e83de4153a568203ea0821) that changed `fftw`. The environments we currently track in CI have been concretized, and each of the specs that might be influenced by this change has been rebuilt in the pipeline.

**Reply by alalazo** [reply]
The environments we currently test in CI are stored in:
- `share/spack/gitlab/cloud_pipelines/stacks/`

**Reply by adrienbernede** [reply]
I don't know how much the Gitlab Pipelines in Spack PRs can grow, and you may want to test on your system.
So, in case the challenge is to setup such testing yourself, you may want to have a look to the work done by:
- NERSC (LBNL): https://docs.nersc.gov/applications/e4s/spack_gitlab_pipeline/
- RADIUSS (LLNL): https://radiuss-spack-testing.readthedocs.io/en/latest/

**Reply by psakievich** [reply]
@alalazo I will need to digest the pipelines a little more to be comprehensive. However, here are the core differences I see: 

- I am asking for a way to test concretization only for specific machine/architecture combinations on an agnostic test machine.
- This is intended to be a light weight option for testing changes in spack that is focused on users rather than spack as a whole

I think per PR is pipelines are not scalable or comprehensive enough for spack.  Consider packages like `Trilinos`. The number of variants and conflicts are a combinatoric nightmare and the dependencies on it will likely exercise a large span of that space even if it is sparse. Also it is not necessarily appropriate for spack to maintain usable packages for every working configuration. Sometimes things will need to be broken and users need to adapt.

Theoretically, I __could__ setup pipelines of my own and test that way but I have two main issues with that.

1. I often don't have permissions or resources to setup a CI pipeline on the machines where I am building
2. The stack that I build can take a long time, and there are several combinations so even if I had permissions the resource request would be non-trivial.

Since the machines/modules/externals change very infrequently compared to spack itself diffs on concretization seem like a pretty good smoke screen to indicate if the current spack commit will build successfully for my active environments, and what changes I can expect.

**Reply by alalazo** [reply]
> I think per PR is pipelines are not scalable or comprehensive enough for spack. Consider packages like Trilinos. The number of variants and conflicts are a combinatoric nightmare and the dependencies on it will likely exercise a large span of that space even if it is sparse. Also it is not necessarily appropriate for spack to maintain usable packages for every working configuration. Sometimes things will need to be broken and users need to adapt.

The point of pipelines is in my opinion to keep a software stack stable and as you note they can be setup at each site independently from Spack's own CI. So, going back to the `trilinos` example: there's no need to test every possible combination of variants. If you test an environment you may have a couple of flavors of `trilinos` installations and you want to make sure that those (and everything in the environment depending on those) continue to work. 

Another point is that it's absolutely fine to make changes to packages that break the previous interface to users (e.g. remove a variant). If that change affects your environment the pipeline build will fail, and if the change is legit you need to modify the environment to let the build pass. In any case you know immediately when the environment breaks and you can decide whether fixing the recipes or updating the environment as you see fit.

**Reply by alalazo** [reply]
@scottwittenburg or @tgamblin can correct me if I am stating something wrong, but I think the functionality you are seeking maps almost 1:1 to the generation job of pipelines. The only difference I see at a high level is that you want to compute what changes by comparing 2 lock files, while pipelines do a similar operation by comparing an existing buildcache index to the newest lock file.

**Reply by psakievich** [reply]
> The point of pipelines is in my opinion to keep a software stack stable and as you note they can be setup at each site independently from Spack's own CI. So, going back to the `trilinos` example: there's no need to test every possible combination of variants. If you test an environment you may have a couple of flavors of `trilinos` installations and you want to make sure that those (and everything in the environment depending on those) continue to work.

Sure that makes sense. I guess I could setup a pipeline on each site/machine I am working on and run that nightly or at some frequency.  I still need to coalesce all that data in a central location so I can monitor the machines.  Does the functionality exist for reporting remote pipeline results? 
I agree with you that the pipeline seems like the ideal way to ensure that the stack is stable.  
We do that to some extent with custom packages in my project to generate nightly testing of packages and upload to CDash, but perhaps switching to a pipeline would be better. I'll have to think about that. There is a lot involved that I won't go into here right now. 
Also does the CI need to use a buildcache for comparison? 

I guess one key difference between what you're describing and my use case is I am less worried about the stability of the stack as a whole in this case and more worried about the build environment's stability.  I would consider my stack (Exawind) fairly volatile since it is still working toward a first release.  However, we are relying on spack to manage the build environment as well as the actual dependency libraries for linking against.  When we get a build error after updating spack the first step is to determine if it was due to a spack package change or a change inside one of our libraries. We live on the bleeding edge of multiple packages so we expect to see build failures semi regularly.

I am open to adopting the pipelines and if that workflow can be adopted to include the strengths of what I'm proposing then I'm all in on that.  I guess the principal things that I don't like about committing to per-machine pipelines in the current state is that it seems like overkill for what I'm after and a lot of extra work to maintain and monitor them.  Theoretically everything spack needs to concretize is contained in a verbose `spack.yaml` file minus things like architecture detection that would need to be overridden. However since this should all live in memory at some point it seems like a reasonable thing to do.

I also think this could have positive impacts on spack's unit-test and PR test procedures.

What I'm suggesting could still be hooked up through a pipeline, use a github action or some other CI interface, but could test a large combination of configurations I actually use on comparatively trivial resources.  For the most part we would do everything a pipeline does minus the actual build and install process through `spack install`.  If we go through the pipeline interface that is perfectly acceptable to me
